### DISCUSSION TITLE
Making `topological_sort` a Graph class method

### DISCUSSION THREAD

**Discussion by adam2392** [networkx/networkx#5419] [Ideas]
Hi,

For the case of supporting multiple edges, I am using a `MultiDiGraph` to represent both `bidirected` and `directed` edges in a causal DAG. One of the core algorithms in causal DAGs is "d-separation". I was thinking I can just write a wrapper for the current `networkx.algorithms.d_separated`, but underneath the hood it calls `topological_sort`. D-separation only runs on directed graphs, so `MultiDiGraph` makes sense, but it can't have cycles. However, the notion of cycles really depends on what kind of edges constitute a cycle. And that really depends on how we define a topological order of a graph, which is a very low-level operation on a graph.

In the interest of enabling networkx as a platform for further development, it would be nice if certain directed core-graph algorithms are class methods inside `DiGraph/MultiDiGraph`, such as `topological_sort`. This `topological_sort` would just call the existing function `topological_sort`, but also enables one to override it in subclasses.

E.g.
```Python
# change 'has_cycle' and functions that call topological_sort to use the class method
def has_cycle(G):
    """Decides whether the directed graph has a cycle."""
    try:
        # Feed the entire iterator into a zero-length deque.
        deque(G.topological_sort()), maxlen=0)
    except nx.NetworkXUnfeasible:
        return True
    else:
        return False

# then in DiGraph
def topological_sort(self, ...):
      # call topological_sort
      topological_sort(...)

# then for ppl that subclass MultiDiGraph and want to use "edge attributes" for representing other directed edges, they can override
def topological_sort(self, ...):
     # only get the directed portion
     dag_g = self._get_directed_portion_of_graph()
     
     # call topological_sort again from networkx
     topological_sort
```

#### Misc Comments
I realize this might be seen as a "large change", but it is the pattern used in many other platform packages, such as MNE-Python (data structure classes call the underlying functions that implement the algorithm, so then subclassing can easily modify certain behavior to call the same function) and scikit-learn (Base classes are very abstract and modularizable).

Plus I believe `topological_sort` only affects a few functions, so it can be easily integrated and all it requires is a refactoring of `nx.topological_sort(G)` to `G.topological_sort()`.

This would begin to enable a larger class of graphs that can be subclassed from the `MultiGraph` class. I think other low-level algorithms are also possible candidates.

**Comment by adam2392** [comment]
I took a look at the codebase inside VScode and only 4 files call `topological_sort(G)`, with the majority being in test files. Thus it seems like a backwards-compatible change that would enable a suite of desired behavior downstream.

**Comment by rossbar** [comment]
I'm -1 on this proposal as it breaks with the high-level design of NetworkX. Generally, NetworkX tries to keep the data structures minimal and extremely flexible, and have a functional interface for operating on graphs. Breaking from this pattern would bloat the data structures and is fraught with the potential for leaky abstractions.

With the flexibility of creating graphs/subgraphs and storing edge attributes, it's very likely that you'll be able to accomplish the analysis you're interested in with the existing data structures and functions, though it might require reframing the problem.

**Comment by dschult** [comment]
Why not just write a function `mixed_graph_topological_sort`?  Whether it gets put inside a class as a method or not doesn't matter.  If you really want it in your subclass, your subclass can create such a method without affecting NetworkX at all. I don't see the advantage of putting lots of methods into the base class.

The disadvantage of putting lots of methods into the base classes is that you have to maintain every method for every single graph class. IMO it's better to house all the code for e.g. topological_sort near the algorithm code rather than near the data structure code.

**Reply by adam2392** [reply]
Okay that makes sense!

I like the idea of `mixed_graph_topological_sort`, where you can then specify the "type of edge" based on the edge attributes. 

Would this function be good to stick into networkx as a PR?

**Reply by dschult** [reply]
If that is the only mixed graph aspect, then I don't think it is universal enough to warrant inclusion. I mean that other people will have different desired edge types and want to specify them in different ways. It could be a good example for the `examples/drawing` section though!!

**Reply by adam2392** [reply]
Wouldn't the general `mixed_graph_topological_sort` be just a wrapper for `topological_sort`, where now there's an additional parameter, denoting the edge attribute? Wouldn't this be a universal function, or am I misunderstanding?

i.e.

```Python
def mixed_graph_topological_sort(G, edge_attr=None):
     # check if edge_attr is in graph
     if edge_attr not in G.edge_attrs:
          raise Error

     # make copy of graph
     G_copy = G.copy()

     # drop all edges that aren't attribute 'edge_attr'
     G_copy.drop_edges(edge_attr)
     
     # call existing topological_sort
     return topological_sort(G_copy)
```
